opendevin/server/session.py

```diff
@@ -89,13 +92,45 @@ async def start_listening(self):
                     await self.create_controller(data)
                 elif action == "start":
                     await self.start_task(data)
+                elif action == "file_selected":
+                    if self.controller is None:
+                        await self.send_error(
+                            "No agent started. Please wait a second..."
+                        )
+                    else:
+                        with open(
+                            Path(self.controller.workdir, data["args"]["file"]), "r"
+                        ) as file:
+                            content = file.read()
+                        await self.send({"action": "file_selected", "message": content})
+                elif action == "refresh_files":
+                    if self.controller is None:
+                        await self.send_error(
+                            "No agent started. Please wait a second..."
+                        )
+                    else:
+                        structure = files.get_folder_structure(
+                            Path(self.controller.workdir)
+                        )
+                        await self.send(
+                            {
+                                "action": "refresh_files",
+                                "message": json.dumps(structure.to_dict()),
+                            }
+                        )
```

Comment:
I think it'd be better to implement these as REST APIs, rather than streaming them through the websocket. I'm OK merging this as-is if it's hard to change, but would probably open up an issue to change it. WDYT?

Reply:
Agreed, wanted to work on it now, this was more of a quick and dirty solution. I guess let me try for an hour or two and if REST is super complicated we can merge this and follow up with a new issue?

Reply:
Aight, should be ready for review again.